### DISCUSSION TITLE
Configuration for GCC packages @ Repology

### DISCUSSION THREAD

**Discussion by ben-foxmoore** [renovatebot/renovate#9686] [Request Help]
I'm struggling to get the right configuration in order to update GCC packages properly, using Repology as the datasource.

I have a YAML file which lists a bunch of APT packages to be installed into an Ubuntu 20.04 based Docker image - it looks like this:
```
packages:
  apt:
    - cmake: 3.16.3-1ubuntu1
    - gcc-9: 9.3.0-17ubuntu1~20.04
    - g++-9: 9.3.0-17ubuntu1~20.04
    - git: 1:2.17.1-1ubuntu0.4
```
and I'm using RegexManager in order to match against this:
```
"regexManagers": [
    {
      "description": "Update APT package versions in packages.yaml",
      "fileMatch": ["(^|/)packages.yaml$"],
      "matchStringsStrategy": "recursive",
      "matchStrings": [
        "\\n(\\s+)apt:\\s*\\n(?:\\s*[#-].*\\n)+",
        "-\\s+(?<depName>\\S*)\\s*:\\s+?(?<currentValue>\\S*)\\s*(#\\s*depName=(?<depNameForced>\\S*))?\\s*\\n"
      ],
      "depNameTemplate": "ubuntu_20_04/{{#if depNameForced}}{{{depNameForced}}}{{else}}{{{depName}}}{{/if}}",
      "datasourceTemplate": "repology",
      "versioningTemplate": "loose"
    }
]
```

---

This works fine for most of the packages, e.g. I've got a PR for bumping package `git` to version `1:2.25.1-1ubuntu3.1` which is correct - here's a screenshot from [Ubuntu's package website](https://packages.ubuntu.com/focal-updates/git): 
![image](https://user-images.githubusercontent.com/5604499/115740339-01c0de80-a38f-11eb-98e5-dbf9223c1caf.png)

Unfortunately, for both `gcc-9` and `g++-9`, I get PRs which aren't accurate. The latest version for `gcc-9` is `9.3.0-17ubuntu1~20.04`:
![image](https://user-images.githubusercontent.com/5604499/115740820-7dbb2680-a38f-11eb-8470-71d1595f5014.png)

However the PR I get tries to update to version `10.2.0-5ubuntu1~20.04` - this is actually the latest version of the `gcc-10` package:
![image](https://user-images.githubusercontent.com/5604499/115741107-bbb84a80-a38f-11eb-8a87-71b7fca06a43.png)

---

I believe this might be related to #8482 (and the associated PR #8489). The query to Repology using the `project-by` tool for `gcc-9` in repo `ubuntu_20_04` gets redirected to `/api/v1/project/gcc` - obviously this matches a who lot of versions of GCC in Ubuntu 20.04 - here's a subset:
```
...
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal/main",
  "srcname": "gcc-9",
  "visiblename": "gcc-9",
  "version": "9.3.0",
  "maintainers": [
    "ubuntu-devel-discuss@lists.ubuntu.com",
    "doko@debian.org"
  ],
  "categories": [
    "devel"
  ],
  "status": "legacy",
  "origversion": "9.3.0-10ubuntu2"
},
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal/universe",
  "srcname": "gcc-8",
  "visiblename": "gcc-8",
  "version": "8.4.0",
  "maintainers": [
    "ubuntu-devel-discuss@lists.ubuntu.com",
    "doko@debian.org"
  ],
  "categories": [
    "universe/devel"
  ],
  "status": "legacy",
  "origversion": "8.4.0-3ubuntu2"
},
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal-updates/main",
  "srcname": "gcc-10",
  "visiblename": "gcc-10",
  "version": "10.2.0",
  "maintainers": [
    "ubuntu-devel-discuss@lists.ubuntu.com",
    "doko@debian.org"
  ],
  "categories": [
    "devel"
  ],
  "status": "outdated",
  "origversion": "10.2.0-5ubuntu1~20.04"
},
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal-updates/main",
  "srcname": "gcc-9",
  "visiblename": "gcc-9",
  "version": "9.3.0",
  "maintainers": [
    "ubuntu-devel-discuss@lists.ubuntu.com",
    "doko@debian.org"
  ],
  "categories": [
    "devel"
  ],
  "status": "legacy",
  "origversion": "9.3.0-17ubuntu1~20.04"
},
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal/main",
  "srcname": "gcc-10",
  "visiblename": "gcc-10",
  "version": "10",
  "maintainers": [
    "ubuntu-devel-discuss@lists.ubuntu.com",
    "doko@debian.org"
  ],
  "categories": [
    "devel"
  ],
  "status": "outdated",
  "origversion": "10-20200411-0ubuntu1"
},
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal/universe",
  "srcname": "gcc-3.3",
  "visiblename": "gcc-3.3",
  "version": "3.3.6ds1",
  "maintainers": [
    "pkern@debian.org"
  ],
  "categories": [
    "universe/devel"
  ],
  "status": "outdated",
  "origversion": "1:3.3.6ds1-30ubuntu2"
},
{
  "repo": "ubuntu_20_04",
  "subrepo": "focal/universe",
  "srcname": "gcc-7",
  "visiblename": "gcc-7",
  "version": "7.5.0",
  "maintainers": [
    "ubuntu-devel-discuss@lists.ubuntu.com",
    "doko@debian.org"
  ],
  "categories": [
    "universe/devel"
  ],
  "status": "legacy",
  "origversion": "7.5.0-6ubuntu2"
},
..
```

The code appears to be trying to match on `binname` and then `srcname`. From what I can see, none of the packages in Repology's response have a `binname` field, but for some reason, it doesn't seem to be filtering on the `srcname`. I guess it's just taking the latest `origversion` that it can find from the list.

Is there anyway to force Renovate to consider the `srcname`, either after `binname` or as a priority? Is there some other option that I'm missing that might help in another way?

Thanks in advance for any input on this issue!

**Comment by HonkingGoose** [comment]
@rarkins can you offer any advice/help here? This discussion is still waiting for a response.

**Reply by rarkins** [reply]
No, unfortunately I have no experience in this area

**Comment by MPV** [comment]
If you're up for trying the `deb` datasource instead of `repology`, I have a working example (including `tzdata`) here:
- https://github.com/MPV/renovate-ubuntu-locales
